### PR TITLE
enable use of lapack by default

### PR DESCRIPTION
**Pull request by asmushetzel**
Changes to makefiles/config to safely enable use of lapack based functionalities by default. By platform:
- Windows: No issues as we get lapack  with openblas automatically
- osx: No issues as we get lapack with "apple" automatically
- Ubuntu: On Ubuntu 14.04 we don't get the lapack-library automatically installed with openblas. On all Ubuntu versions we don't get it automatically installed when using "atlas". The simplest and most consistent handling is to explicitly check whether the lapack-library is in /usr/lib or /usr/local/lib and if not, then switch off lapack support automatically.

In addition provided an extra variable in the config to support setting a custom path for the lapack library. Just for convenience. 

BTW: The build using "atlas" is generally broken on Ubuntu (at least). Reason is that the mshadow-config adds a "-lcblas" while Ubuntu will install a "blas"-library. The latter is in contrast to the official atlas documentation which states that the library is "cblas". But as atlas is not providing precompiled libraries, guess anyone else is free to name them differently and apparently that is what happened. We have no regression build for atlas on Jenkins, so we never see this. 
Eric, do you think we can simply change this setting in mshadow?

### CODE REVIEW

**Comment by asmushetzel** [comment]
@jli05

File: make/config.mk [hunk 1]
```diff
@@ -97,6 +93,32 @@ else
...
+# path to lapack library in case of a non-standard installation
+USE_LAPACK_PATH =
+
+# verify existence of separate lapack library
+# switch off lapack support in case it can't be found
+# issue covered with this 
+#   -  for Ubuntu 14.04 or lower, lapack is not automatically installed with openblas 
+#   -  for Ubuntu, installing atlas will not automatically install the atlas provided lapack library
+# silently switching lapack off instead of letting the build fail because of backward compatibility
+ifeq ($(USE_LAPACK), 1)
```
Comment (piiswrong) [inline make/config.mk]: Move this to Makefile

File: make/config.mk [hunk 2]
```diff
@@ -97,6 +93,32 @@ else
 USE_BLAS = atlas
 endif
 
+# whether use lapack during compilation
+# only effective when compiled with blas versions openblas/apple/atlas/mkl
+USE_LAPACK = 1
+
+# Name of separate lapack library
+LAPACK_LIB = liblapack.a
```
Comment (piiswrong) [inline make/config.mk]: whats the purpose of this?

Comment (asmushetzel) [inline make/config.mk]: About the LAPACK_LIB? I need the name of the library that must be present when looking into the subdirectories the lines below. Or do you mean something else?

**Comment by piiswrong** [comment]
Looks like the linux-cpu test machine doesn't have lapack installed


terminate called after throwing an instance of 'dmlc::Error'
  what():  [13:55:17] src/engine/./threaded_engine.h:329: [13:55:17] include/mxnet/c_lapack_api.h:84: MXNet build without lapack. Function spotrf is not available.

**Comment by piiswrong** [comment]
The test passed before, so it's probably because its not detected?

http://ec2-52-25-96-65.us-west-2.compute.amazonaws.com/blue/organizations/jenkins/mxnet/detail/PR-6538/22/pipeline/221

**Comment by asmushetzel** [comment]
Debugging this. Don't can really see the full output of the compile step if this passed but the test failed. Makes it difficult to debug. But will find it

**Comment by asmushetzel** [comment]
All resolved now.

**Comment by piiswrong** [comment]
Thanks!

**Comment by liangfu** [comment]
The compiler raises error at link stage when ```lapack.so``` file exist at ```/usr/lib/lapack/lapack.so``` (the default ```lapack.so``` path in Debian 9) instead of ```/usr/lib/lapack.so```. I think the search path of ```lapack.so```should be localized by cmake instead of in the Makefile.

**Comment by asmushetzel** [comment]
Do I understand it correctly that you are using cmake to build on debian instead of plain make?

**Comment by liangfu** [comment]
Yes, it's true. and i was confused. But cmake failed to localize the lapack.so file, and since there seems incremental compilation disabled in mxnet, I have to compile all over again.

**Comment by szha** [comment]
@asmushetzel what's your suggestion on mac? I've been maintaining pip prebuilt binaries, which is currently using openblas and it does have lapack, but due to its dependency on fortran compiler (not shipped with Xcode CLT), the build process has been very painful and unsuccessful. I've tried using gfortran from homebrew but it ended up with duplicate symbol conflicts that I can't resolve. Static linking is preferred, and dynamic linking should be limited to libraries that are shipped with the OS.

**Comment by jli05** [comment]
Shouldn't the macOS prebuilt binary be using Accelerate/vecLib by default?

**Comment by jli05** [comment]
Is OpenBLAS's support for macOS good/mature?

**Comment by asmushetzel** [comment]
The config we provide with MXNet for osx is based on accelerate framework and that has all lapack in there and works. If you instead use OpenBlas on mac, you  either have to go through the process of compiling the lapack libraries as well or switch off use of lapack in the config-file.

**Comment by liangfu** [comment]
Is there any side effect by defining USE_LAPACK=0, and why USE_LAPACK is defined to be 1 by default more recently?

**Comment by asmushetzel** [comment]
You can't use some linalg-operators without it. Otherwise no side effect. It seems as if you are doing a custom installation but anyway it should then switch itself off automatically. Can you supply the detailed scripts that you used for your make process (in case that you have edited cmakefile etc). Then I can take a look whether there is something to improve.

**Comment by szha** [comment]
@jli05 the cblas portion of openblas has been ok.

**Comment by szha** [comment]
@jli05 @asmushetzel thanks for the comments. I'm switching the pre-built binary to the accelerate framework going forward until openblas has better support on mac for lapack, at which time I will re-evaluate the decision based on performance.

**Pull request merged**